Socket: improve cross-platform behavior on Dispose

tmds opened the pull request:
- Unix: unblock synchronous calls on `Dispose`
- Windows: implement unblock using `CancelIoEx`
- Unix/Windows: uses `SafeHandle` instead of `IntPtr` (needed on Windows for `CancelIoEx` unblocking)
- Unix: match when Windows sends RST on Close
- Unix/Windows: same `SocketError` for aborted operations

based on https://github.com/dotnet/corefx/pull/37486

fixes #22564
fixes #26034

CC @wfurt @stephentoub @geoffkizer @davidsh @krwq

tmds commented:
I'm first letting the tests of https://github.com/dotnet/corefx/pull/37486 run on CI to ensure they are asserting current corefx behaviour on Windows.

tmds commented:
> I'm first letting the tests of #37486 run on CI to ensure they are asserting current corefx behaviour on Windows.

CI passes on Windows for all test except `SyncSendFileGetsCanceledByDispose`. For that test it fails because the operation isn't aborting on Dispose.

davidsh commented:
>CI passes on Windows for all test except SyncSendFileGetsCanceledByDispose. For that test it fails because the operation isn't aborting on Dispose.

Most of the CI test failures that we saw with the original PR were from Outerloop.

davidsh commented:
/azp run

davidsh commented:
/azp run

tmds commented:
Based on reported issues for the previous PR this has changed:
- Cleanup the SO_TYPE socket length handling https://github.com/dotnet/corefx/pull/38804/commits/a1f790c10064fce8d5f616f21d9aa675a16e11bc
- Update SocketErrors when Cleanup for APM APIs https://github.com/dotnet/corefx/pull/38804/commits/387542e749a2b165d8f4a8030204d92802e73303. This was causing unexpected SocketErrors for canceled operations.

tmds commented:
CI failures:

- ` System.Tests.GetCommandLineArgs/GetCommandLineArgs_Invoke_ReturnsExpected(args: [\"\\\"Arg1 With Quotes\\\"\", \"\\\"Arg2 With Quotes\\\"\"])` failed on Windows nano:
```
Exit code was -1073740940 but it should have been 42
Expected: True
Actual:   False
```
- Catastrophic failure on Windows.7.Amd64.Open-x64 for System.Net.Http.Functional.Tests:
```
Process terminated. Assertion failed.
   at Interop.WinHttp.SafeWinHttpHandle.SetParentHandle(SafeWinHttpHandle parentHandle) in /_/src/Common/src/Interop/Windows/WinHttp/Interop.SafeWinHttpHandle.cs:line 35

   at System.Net.Http.WinHttpHandler.StartRequestAsync(WinHttpRequestState state) in /_/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:line 813

   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in /_/src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:line 1019

   at System.Net.Http.WinHttpHandler.StartRequestAsync(WinHttpRequestState state)

   at System.Net.Http.WinHttpHandler.<>c.<SendAsync>b__101_0(Object s) in /_/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs:line 560

   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:line 289

   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2389

   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /_/src/System.Private.CoreLib/shared/System/Threading/ThreadPool.cs:line 663
----- end Tue 06/25/2019  5:07:26.86 ----- exit code -2146232797 ----------------------------------------------------------
```
- mc.dot.net isn't showing results for `Windows.10.Amd64.ClientRS5`.

tmds commented:
@davidsh @stephentoub @krwq @wfurt maybe we can merge this after a few more CI runs?

stephentoub commented on src/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -2618,6 +2667,32 @@ void SystemNative_GetDomainSocketSizes(int32_t* pathOffset, int32_t* pathSize, i
+    err = connect(fd, &addr, sizeof(addr));
+#elif HAVE_DISCONNECTX
+    // disconnectx causes a FIN close on OSX. It's the best we can do.
+    err = disconnectx(fd, SAE_ASSOCID_ANY, SAE_CONNID_ANY);
```
Is this going to change existing behavior on macOS?

tmds replied:
With this, macOS gets unblocked, but compared to Windows&Linux, the peer sees a FIN close instead of RST close.

stephentoub commented on src/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -2618,6 +2667,32 @@ void SystemNative_GetDomainSocketSizes(int32_t* pathOffset, int32_t* pathSize, i
+
+#if defined(__linux__)
+    // On Linux, we can disconnect a socket by connecting to AF_UNSPEC.
+    // For TCP sockets, this causes an abortive close.
```
What happens if this is used on connectionless sockets?

tmds replied:
It will give back the local address, so it may be used by another socket. The peer doesn't observe something.

stephentoub commented on src/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -2618,6 +2667,32 @@ void SystemNative_GetDomainSocketSizes(int32_t* pathOffset, int32_t* pathSize, i
+    // On Linux, we can disconnect a socket by connecting to AF_UNSPEC.
+    // For TCP sockets, this causes an abortive close.
+
+    // We don't need to clear the address, only the sa_family will be used.
```
I know we spoke about this, but I'd still prefer it if the struct were cleared, rather than relying on an implementation that the address will be ignored.

tmds replied:
I'll clear it.

stephentoub commented on src/Native/Unix/System.Native/pal_networking.c:
```diff
@@ -2618,6 +2667,32 @@ void SystemNative_GetDomainSocketSizes(int32_t* pathOffset, int32_t* pathSize, i
+#elif HAVE_DISCONNECTX
+    // disconnectx causes a FIN close on OSX. It's the best we can do.
+    err = disconnectx(fd, SAE_ASSOCID_ANY, SAE_CONNID_ANY);
+#else
```
shutdown can't be used here?  It's not reliable?  It doesn't have the desired behavior?

tmds replied:
I'm fine using it as a fall-back. At best it will cause a FIN close.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/AcceptOverlappedAsyncResult.Windows.cs:
```diff
@@ -33,8 +33,12 @@ internal override object PostCompletion(int numBytes)
                 try
                 {
+                    _listenSocket.SafeHandle.DangerousAddRef(ref refAdded);
+                    IntPtr handle = _listenSocket.SafeHandle.DangerousGetHandle();
```
Nit: I'd feel better about this if _listenSocket.SafeHandle were first stored into a local, and then that local used for all three of the DangerousAddRef/GetHandle/Release calls.  It's not obvious that _listenSocket._handle can't change concurrently between the first and last access.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -195,12 +196,16 @@ internal static unsafe SocketError Accept(SafeSocketHandle socketHandle, byte[]
             }
+
+            return aborted;
         }
```
Nit: this could be simplified to something like:
```C#
private bool InnerReleaseHandle() => _asyncContext?.StopAndAbort() ?? false;
```

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -243,10 +248,10 @@ private unsafe SocketError InnerReleaseHandle()
-                if (_blockable)
+                if (!_abortive)
                 {
                     if (NetEventSource.IsEnabled) NetEventSource.Info(this, $"handle:{handle} Following 'blockable' branch.");
```
This message is no longer accurate.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -268,7 +273,7 @@ private unsafe SocketError InnerReleaseHandle()
                 }
 
-                // By default or if CloseAsIs() path failed, set linger timeout to zero to get an abortive close (RST).
+                // By default or non abortive path failed, set linger timeout to zero to get an abortive close (RST).
```
Nit: "or non abortive path" => "or if the non-abortive path"

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -358,6 +363,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
+            internal unsafe bool TryUnblockSocket()
+            {
+                // Calling 'close' on a socket that has pending blocking calls (e.g. recv, send, accept, ...)
+                // may block indefinitly. This is a best effort attempt to not get blocked and make those operations return.
```
Nit: indefinitly => indefinitely

stephentoub replied:
Nit: best effort => best-effort

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -358,6 +363,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
+                // What we do here isn't specified by POSIX and doesn't work on all OSes.
+                // On Linux this works well.
+                // On OSX, TCP connections will be closed with a FIN close instead of an abortive RST close.
+                // And, pending TCP connect operations and UDP receive are not abortable.
```
Is this going to make things worse than they are today?  With this change, Dispose'ing of the associated SafeHandle will never unblock a synchronous Close/Accept, because the SafeHandle will have been add-ref'd.  Might it have before this change, even if it was racy?

tmds replied:
I think there are no regressions. I've opened https://github.com/dotnet/corefx/pull/39079 to check how tests behaved before Socket changes.

jkotalik replied:
We are seeing some issues with this change in Kestrel in the following test: https://github.com/aspnet/AspNetCore/blob/3a5dbad3bcf66750000943194f7270a2df3cc33d/src/Servers/Kestrel/test/FunctionalTests/ResponseTests.cs#L905. This is only failing on linux. 

I'm not sure what specifically in this change caused the regression, but we saw this test fail ~30% of the time in Kestrel. More information here: https://github.com/aspnet/AspNetCore/pull/13532

Here are the Kestrel logs we are seeing:
```
| [0.001s] TestLifetime Information: Starting test ClientCanReceiveFullConnectionCloseResponseWithoutErrorAtALowDataRate at 2019-09-09T09:05:06
| [0.068s] Microsoft.AspNetCore.Hosting.Diagnostics Debug: Hosting starting
| [0.071s] Microsoft.AspNetCore.Hosting.Diagnostics Debug: Hosting started
| [0.071s] Microsoft.AspNetCore.Hosting.Diagnostics Debug: Loaded hosting startup assembly Sockets.FunctionalTests, Version=42.42.42.42, Culture=neutral, PublicKeyToken=adb9793829ddae60
| [0.072s] Microsoft.AspNetCore.Server.Kestrel Debug: TestServer is listening on port 34479
| [0.074s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLPL0J6HPQ7C" accepted.
| [0.077s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLPL0J6HPQ7C" started.
| [0.078s] Microsoft.AspNetCore.Hosting.Diagnostics Information: Request starting HTTP/1.1 GET http:/// - -
| [11.186s] Microsoft.AspNetCore.Hosting.Diagnostics Information: Request finished HTTP/1.1 GET http:/// - - - 200 - - 11107.8383ms
| [11.187s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLPL0J6HPQ7C" disconnecting.
| [11.188s] Microsoft.AspNetCore.Server.Kestrel Debug: Connection id "0HLPL0J6HPQ7C" stopped.
| [11.188s] Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets Debug: Connection id "0HLPL0J6HPQ7C" sending FIN because: "The Socket transport's send loop completed gracefully."
| [14.193s] Microsoft.AspNetCore.Hosting.Diagnostics Debug: Hosting shutdown
| [14.886s] TestLifetime Information: Finished test ClientCanReceiveFullConnectionCloseResponseWithoutErrorAtALowDataRate in 14.885012699999999s
```

This also could be an issue with how we are handling socket shutdown. Here is the shutdown code in Kestrel: https://github.com/aspnet/AspNetCore/blob/master/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L337

@tmds can you give me a TL;DR on what changed here in Socket.Shutdown + Socket.Dispose on Unix? 

cc @halter73

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -358,6 +363,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
+                if (fdFlags == 0)
+                {
+                    return false;
+                }
```
Same question here.  Is this possibly making things worse?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs:
```diff
@@ -358,6 +363,43 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
+                int type = 0;
+                int optLen = sizeof(int);
+                Interop.Error err = Interop.Sys.GetSockOpt(this, SocketOptionLevel.Socket, SocketOptionName.Type, (byte*)&type, &optLen);
+                Debug.Assert(err == Interop.Error.SUCCESS);
```
In what situation might this fail?  If it does fail, what will our follow-up action be... remove the assert?  We generally don't assert on the results of P/Invokes like this... what makes this one ok?

tmds replied:
I'll remove this.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs:
```diff
@@ -108,14 +108,16 @@ internal static SafeSocketHandle CreateWSASocket(AddressFamily addressFamily, So
                 _iocpBoundHandle.Dispose();
             }
+
+            return false;
```
What does `return false` mean here?  Generally the result of a "ReleaseHandle" call is true for success and false for catastrophic failure.

tmds replied:
The return value indicates whether the connection should do an abortive close. I'm going to rename this method to `DoReleaseHandle` and add some comments.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs:
```diff
@@ -222,13 +224,19 @@ internal static InnerSafeCloseSocket CreateWSASocket(AddressFamily addressFamily
+            internal unsafe bool TryUnblockSocket()
+            {
+                // Try to cancel all pending IO.
+                return Interop.Kernel32.CancelIoEx(this, null);
```
Remind me what makes the whole operation thread-safe with regards to race conditions that would cause an additional operation to be created after this call to CancelIoEx (in which case that additional operation wouldn't be canceled by this CancelIoEx and would remain blocked)?  There's something somewhere that ensures that by the time we make this call, no additional operations can be created?

tmds replied:
Starting a new operation will throw ODE from the Socket class.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -5225,6 +5271,59 @@ private void ValidateBlockingMode()
+            }
+        }
+
+        private SocketError UpdateSendSocketErrorForCleanedUp(SocketError socketErrorIn)
```
Every use of this is:
```C#
errorCode = UpdateSendSocketErrorForCleanedUp(errorCode);
```
Should this instead return void and take a ref SocketError?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -5225,6 +5271,59 @@ private void ValidateBlockingMode()
+            }
+        }
+
+        private SocketError UpdateConnectSocketErrorForCleanedUp(SocketError socketErrorIn)
```
Ditto

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -5225,6 +5271,59 @@ private void ValidateBlockingMode()
+            }
+        }
+
+        private SocketError UpdateAcceptSocketErrorForCleanedUp(SocketError socketErrorIn)
```
Ditto

davidsh commented:
>@davidsh @stephentoub @krwq @wfurt maybe we can merge this after a few more CI runs?

There needs to be a lot more Outerloop CI runs before we can understand the stability of this. Everytime a commit is pushed only Innerloop CI is run. 

Outerloop CI needs to be run manually after every commit push with "/azp run".

davidsh commented:
/azp run

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs:
```diff
@@ -177,7 +179,7 @@ private SocketError InnerReleaseHandle()
                 }
 
-                // By default or if CloseAsIs() path failed, set linger timeout to zero to get an abortive close (RST).
+                // By default or if non abortive path failed, set linger timeout to zero to get an abortive close (RST).
```
Nit: non-abortive

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs:
```diff
@@ -197,7 +201,7 @@ internal sealed partial class InnerSafeCloseSocket : SafeHandleMinusOneIsInvalid
             private InnerSafeCloseSocket() : base(true) { }
 
-            private bool _blockable;
+            private bool _abortive = true;
```
Can you add a comment about what "abortive" means here?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -1433,6 +1442,8 @@ public int Receive(byte[] buffer, int offset, int size, SocketFlags socketFlags,
             int bytesTransferred;
             errorCode = SocketPal.Receive(_handle, buffer, offset, size, socketFlags, out bytesTransferred);
 
+            errorCode = UpdateReceiveSocketErrorForCleanedUp(errorCode, bytesTransferred);
```
Will this ever change from a success error code to non-success, or vice versa?  Assuming not, can it be moved off of the success path into the below if block?

tmds replied:
Yes, for UDP sockets on Linux, when we abort the operation, bytesReceived returns zero with Success. UpdateReceiveSocketErrorForCleanedUp includes a comment about this.

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -1480,6 +1491,8 @@ public int Receive(Span<byte> buffer, SocketFlags socketFlags, out SocketError e
             int bytesTransferred;
             errorCode = SocketPal.Receive(_handle, buffer, socketFlags, out bytesTransferred);
 
+            errorCode = UpdateReceiveSocketErrorForCleanedUp(errorCode, bytesTransferred);
```
Ditto for all of these

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -228,10 +228,9 @@ public bool TryCancel()
                             // The previous canceller will have fired the completion, etc.
                             Trace("Exit, previously cancelled");
-                            return true;
+                            return false;
```
Where is this result used?  I'm wondering why TryCancel doesn't just return `void`.

stephentoub replied:
I see, you added a use for it below.

stephentoub replied:
Can you add a descriptive comment to TryCancel indicating what the return value means?

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:
```diff
@@ -736,7 +735,7 @@ public bool IsReady(SocketAsyncContext context, out int observedSequenceNumber)
                 {
                     observedSequenceNumber = _sequenceNumber;
-                    bool isReady = (_state == QueueState.Ready);
+                    bool isReady = (_state == QueueState.Ready) || (_state == QueueState.Stopped);
```
Why did this change?

tmds replied:
I got a test failure because this didn't cause an operation to be tried synchronously when the queue was stopped. It failed somewhere later, I believe in a sync-over-async path.
By returning true for stopped, this will throw ODE when the operation is performed synchronously.

tmds replied:
I got the same test failure with https://github.com/dotnet/corefx/pull/39079: 
```
Process terminated. Assertion failed.
   at System.Net.Sockets.SocketAsyncContext.Register() in /_/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:line 1150
   at System.Net.Sockets.SocketAsyncContext.OperationQueue`1.StartAsyncOperation(SocketAsyncContext context, TOperation operation, Int32 observedSequenceNumber, CancellationToken cancellationToken) in /_/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:line 754
   at System.Net.Sockets.SocketAsyncContext.PerformSyncOperation[TOperation](OperationQueue`1& queue, TOperation operation, Int32 timeout, Int32 observedSequenceNumber) in /_/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:line 1225
   at System.Net.Sockets.SocketAsyncContext.SendFile(SafeFileHandle fileHandle, Int64 offset, Int64 count, Int32 timeout, Int64& bytesSent) in /_/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs:line 1905
   at System.Net.Sockets.SocketPal.SendFile(SafeSocketHandle handle, FileStream fileStream) in /_/src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Unix.cs:line 933
   at System.Net.Sockets.Socket.SendFileInternal(String fileName, Byte[] preBuffer, Byte[] postBuffer, TransmitFileOptions flags) in /_/src/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs:line 148
   at System.Net.Sockets.Socket.SendFile(String fileName, Byte[] preBuffer, Byte[] postBuffer, TransmitFileOptions flags) in /_/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 1292
   at System.Net.Sockets.Socket.SendFile(String fileName) in /_/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:line 1271
   at System.Net.Sockets.Tests.SendFileTest.<>c__DisplayClass7_0.<SyncSendFileGetsCanceledByDispose>b__0() in /_/src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs:line 209
   at System.Threading.Tasks.Task.InnerInvoke() in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2449
   at System.Threading.Tasks.Task.<>c.<.cctor>b__274_0(Object obj) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2428
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:line 172
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2385
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(Thread threadPoolThread) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/Task.cs:line 2327
   at System.Threading.Tasks.ThreadPoolTaskScheduler.<>c.<.cctor>b__10_0(Object s) in /_/src/System.Private.CoreLib/shared/System/Threading/Tasks/ThreadPoolTaskScheduler.cs:line 37
   at System.Threading.ThreadHelper.ThreadStart_Context(Object state) in /_/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:line 50
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in /_/src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:line 172
   at System.Threading.ThreadHelper.ThreadStart(Object obj) in /_/src/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs:line 83
```

stephentoub commented on src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs:
```diff
@@ -1107,8 +1107,12 @@ private unsafe SocketError FinishOperationAccept(Internals.SocketAddress remoteS
             try
             {
+                _currentSocket.SafeHandle.DangerousAddRef(ref refAdded);
+                IntPtr handle = _currentSocket.SafeHandle.DangerousGetHandle();
```
Same comment as earlier about storing the SafeHandle into a local first and operating on that local.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Accept.cs:
```diff
@@ -288,6 +288,75 @@ public async Task AcceptAsync_MultipleAcceptsThenDispose_AcceptsThrowAfterDispos
+
+                Task timeoutTask = Task.Delay(30000);
+
+                Assert.NotSame(timeoutTask, await Task.WhenAny(disposeTask, acceptTask, timeoutTask));
```
This can instead use the WhenAllOrAnyFailed helper, e.g.
```C#
await new Task[] { acceptTask, disposeTask }.WhenAllOrAnyFailed(millisecondsTimeout: 30_000);
```

stephentoub replied:
Ah, you actually don't want to propagate exceptions from the completed task here.  Regardless, if you use Task.Delay for timeouts like this, it's best to cancel them... otherwise they leave pending timers that will eventually fire later, e.g.
```C#
var cts = new CancellationTokenSource();
Task timeoutTask = Task.Delay(30_000, cts.Token);
...
cts.Cancel();
```

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Accept.cs:
```diff
@@ -288,6 +288,75 @@ public async Task AcceptAsync_MultipleAcceptsThenDispose_AcceptsThrowAfterDispos
+            }
+            else
+            {
+                if (UsesSync)
```
Nit: This needn't be nested, e.g.
```C#
if (UsesApm)
{
    ...
}
else if (UsesSync)
{
    ...
}
else
{
    ...
}
```

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Connect.cs:
```diff
@@ -85,6 +85,82 @@ public async Task Connect_AfterDisconnect_Fails()
+                Task connectTask = Task.Factory.StartNew(() =>
+                {
+                    ConnectAsync(client, new IPEndPoint(IPAddress.Parse("1.1.1.1"), 23)).GetAwaiter().GetResult();
+                }, TaskCreationOptions.LongRunning);
```
Why is this spinning up a new thread only to then block it by creating an asynchronous operation and synchronously waiting for it?

tmds replied:
I was a bit too paranoid about blocking the threadpool when writing these tests.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Connect.cs:
```diff
@@ -85,6 +85,82 @@ public async Task Connect_AfterDisconnect_Fails()
+                }, TaskCreationOptions.LongRunning);
+
+                // Wait a little so the operation is started, then Dispose.
+                await Task.Delay(100);
```
Why isn't it sufficient to do:
```C#
Task connectTask = ConnectAsync(...);
... // operation has started here, no delay needed
```
Is the additional delay necessary only for the synchronous case, where we're explicitly queueing a work item to start the operation?  And if so, could we add this delay only for the synchronous case?

tmds replied:
> Is the additional delay necessary only for the synchronous case, where we're explicitly queueing a work item to start the operation? 

Yes, we want the work item to be blocked.

> And if so, could we add this delay only for the synchronous case?

ok

tmds replied:
> And if so, could we add this delay only for the synchronous case?

This is also needed for the async case, otherwise those operations throw ODE and not the SE we are checking.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Connect.cs:
```diff
@@ -85,6 +85,82 @@ public async Task Connect_AfterDisconnect_Fails()
+
+            SocketError? localSocketError = null;
+            bool disposedException = false;
+            for (int i = 0; i < 10 && !localSocketError.HasValue; i++)
```
We have a RetryHelper.  Would that work here rather than a custom loop?

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Connect.cs:
```diff
@@ -85,6 +85,82 @@ public async Task Connect_AfterDisconnect_Fails()
+                {
+                    break;
+                }
+            }
```
Nit: missing blank line after the brace

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/Connect.cs:
```diff
@@ -85,6 +85,82 @@ public async Task Connect_AfterDisconnect_Fails()
+                Assert.Null(localSocketError);
+                Assert.True(disposedException);
+            }
+            else
```
Nit: same unnecessary nesting comment

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs:
```diff
@@ -180,6 +180,100 @@ public void SendFile_Synchronous(IPAddress listenAt, bool sendPreAndPostBuffers,
+                    Task socketOperation = Task.Factory.StartNew(() =>
+                    {
+                        // Create a large file that will cause SendFile to block until the peer starts reading.
+                        string filename = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
```
Can we use FileCleanupTestBase instead?  We should be able to put it at the bottom of the class hierarchy.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs:
```diff
@@ -180,6 +180,100 @@ public void SendFile_Synchronous(IPAddress listenAt, bool sendPreAndPostBuffers,
+                }
+
+                // On OSX, we're unable to unblock the on-going socket operations and
+                // perform an abortive close.
```
Does that mean this test is going to leave blocked sockets?

tmds replied:
No, there are no blocked threads. The peer sees a FIN close on macOS instead of a RST close when we unblock the on-going socket operations.

stephentoub commented on src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs:
```diff
@@ -949,6 +949,179 @@ protected static (Socket, Socket) CreateConnectedSocketPair()
+                Task receiveTask = Task.Factory.StartNew(() =>
+                {
+                    ReceiveAsync(socket, new ArraySegment<byte>(new byte[1])).GetAwaiter().GetResult();
+                }, TaskCreationOptions.LongRunning);
```
Same question about why we're doing it this way.

stephentoub commented:
Thanks, @tmds.

I'm excited to get the change in, as it helps to address a real reliability pain point developers have hit.  At the same time, the scope of the change at this late in the game for .NET Core 3.0 makes me nervous.  I want us to get as much validation as possible before we consider merging this.  That not only includes multiple successful outerloop CI runs, but also validating this with ASP.NET's test suites, both functional and performance.

@sebastienros, @davidfowl, what's the best way to do that?

stephentoub commented:
@geoffkizer, could you review this PR as well?

davidfowl commented:
We currently have a 10% regression that we’re chasing down. So I’d wait on this if we’re. Worried it could regress even more

tmds commented:
@stephentoub thank you for reviewing! I've addressed comments. For two open questions:

> // On OSX, TCP connections will be closed with a FIN close instead of an abortive RST close.
// And, pending TCP connect operations and UDP receive are not abortable.

I ran tests of the PR without the other changes in https://github.com/dotnet/corefx/pull/38804. The tests I'm most interested in (ConnectGetsCanceledByDispose, UdpReceiveGetsCanceledByDispose on macOS) show neither a pass/fail status: https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F39079~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190702.1/workItem/System.Net.Sockets.Tests. I'm not sure what the CI result means then, but it was not passing.

> if (fdFlags == 0)
{
    return false;
}

This is to avoid changing non-CLOEXEC sockets that may be shared with other Processes. I've augmented this to skip the check when the close is abortive (like when the Dispose happens from the finalizer).

stephentoub commented:
> I'm not sure what the CI result means then, but it was not passing

Actually I believe it was. For infrastructure reasons, passing data is at least temporarily not being stored. And if you look at the infrastructure log, you'll see the full xunit output, which shows:
```
===========================================================================================================
~/dotnetbuild/work/3be7edae-df7d-49bf-844e-5d3896a7ea71/Work/61c25c17-c2c2-4d21-a0c3-d09c26840d10/Exec ~/dotnetbuild/work/3be7edae-df7d-49bf-844e-5d3896a7ea71/Work/61c25c17-c2c2-4d21-a0c3-d09c26840d10/Exec
  Discovering: System.Net.Sockets.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Net.Sockets.Tests (found 715 of 1089 test cases)
  Starting:    System.Net.Sockets.Tests (parallel test collections = on, max threads = 8)
    System.Net.Sockets.Tests.UnixDomainSocketTest.Socket_CreateUnixDomainSocket_Throws_OnWindows [SKIP]
      Condition(s) not met: "IsSubWindows10"
    System.Net.Sockets.Tests.KeepAliveTest.Socket_KeepAlive_RetryCount_Failure [SKIP]
      Condition(s) not met: "IsWindowsBelow1703"
    System.Net.Sockets.Tests.CreateSocket.Ctor_Raw_Supported_Success [SKIP]
      Condition(s) not met: "SupportsRawSockets"
    System.Net.Sockets.Tests.SendReceiveMemoryNativeTask.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveTask.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveApm.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveMemoryArrayTask.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveEap.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveMemoryNativeTask.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveTask.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveApm.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveEap.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveMemoryArrayTask.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionReset
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1106,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendFileTest.SyncSendFileGetsCanceledByDispose [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs(230,0): at System.Net.Sockets.Tests.SendFileTest.SyncSendFileGetsCanceledByDispose()
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1061,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.AcceptSyncForceNonBlocking.AcceptGetsCanceledByDispose [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/Accept.cs(320,0): at System.Net.Sockets.Tests.Accept`1.AcceptGetsCanceledByDispose()
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.AcceptSync.AcceptGetsCanceledByDispose [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/Accept.cs(320,0): at System.Net.Sockets.Tests.Accept`1.AcceptGetsCanceledByDispose()
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1061,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1061,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: True) [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1061,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSpanSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionAborted
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1117,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionAborted
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1117,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSyncForceNonBlocking.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionAborted
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1117,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
    System.Net.Sockets.Tests.SendReceiveSpanSync.TcpReceiveSendGetsCanceledByDispose(receiveOrSend: False) [FAIL]
      Assert.Equal() Failure
      Expected: ConnectionAborted
      Actual:   (null)
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(1117,0): at System.Net.Sockets.Tests.SendReceive`1.TcpReceiveSendGetsCanceledByDispose(Boolean receiveOrSend)
        --- End of stack trace from previous location where exception was thrown ---
  Finished:    System.Net.Sockets.Tests
=== TEST EXECUTION SUMMARY ===
   System.Net.Sockets.Tests  Total: 927, Errors: 0, Failed: 21, Skipped: 3, Time: 93.568s
```

tmds commented:
> Actually I believe it was. For infrastructure reasons, passing data is at least temporarily not being stored. And if you look at the infrastructure log, you'll see the full xunit output, which shows:

Thanks. The Connect and UdpReceive tests were skipped on macOS. I've included them now and another CI build is in progress: https://github.com/dotnet/corefx/pull/39079.

tmds commented:
No regression on macOS for {Connect,UdpReceive}GetsCanceledByDispose. Pre-PR:
```
    System.Net.Sockets.Tests.ConnectSync.ConnectGetsCanceledByDispose [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/Connect.cs(118,0): at System.Net.Sockets.Tests.Connect`1.ConnectGetsCanceledByDispose()

    System.Net.Sockets.Tests.SendReceiveSync.UdpReceiveGetsCanceledByDispose [FAIL]
      Assert.NotSame() Failure
      Stack Trace:
        /_/src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs(979,0): at System.Net.Sockets.Tests.SendReceive`1.UdpReceiveGetsCanceledByDispose()
```

tmds commented:
@geoffkizer @davidfowl can you please look into this PR? A previous attempt was reverted after being merged just before preview 7. It would be good if this gets some more time on master.

stephentoub commented:
Thanks, Tom. At this point I think it's going to miss 3.0, and we'll check it in first thing after master branches. The bar is high right now, and we just don't have any more runway to deal with fallout; while I really want this change in, it has the potential to be destabilizing, have unintended perf regressions, etc.
